torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -591,9 +588,22 @@ def _update_trainable(self) -> None:
         # Create the optim which will work on the param shard
         if not hasattr(self, "optim"):
             self._clear_cache()
-            self._default_device = list(self._per_device_params.keys())[0]
             self.optim = self._optim_constructor(self.partition_parameters()[self.rank], **self._optim_defaults)
             self._sync_param_groups(self.optim.param_groups, self.param_groups)
 
         if self.parameters_as_bucket_view:
             self._setup_flat_buffers()
+
+    def _verify_spsd_integrity(self) -> None:
```

Comment:
Good catch! This actually suggests that we should add a followup PR to support multi-device models, where there is still just one model replica per DDP-process, but the model can span multiple GPUs. See this tutorial: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#combine-ddp-with-model-parallelism

Reply:
I realized in regard to supporting this model parallelism: Does this not revert the situation back to the existing code (with respect to `self._per_device_params` and `self.buckets`)?

The existence of `self._per_device_params` and the usage of `self.buckets` as a `Dict[torch.device, List[torch.Tensor]]` rather than a `List[torch.Tensor]` were to accommodate having the model parameters of a single model replica be spread across multiple devices. 

In terms of `self._per_device_params` (of type `Dict[torch.device, List[List[Parameter]]`), the difference between 'SPMD' and 'SPSD with model parallelism' is that in SPMD, for each device, each model parameter appears once in the mapped-to value, while in SPSD with model parallelism, the mapped-to values form a partition of the model parameters. Either way, the layer of indirection through the device is needed.

Reply:
Resolved after discussing with @mrshenli. The current change should stay as is; there is no need for the explicit device layer of indirection even for model parallelism.